### PR TITLE
Fix slow repo_list with many refs

### LINKED ISSUE #309: large repository makes index page very slow
# Issue

Similar to #230 I too have cloned the [nixpkgs](https://github.com/NixOS/nixpkgs/) repository and am now experiencing long wait times.
The corresponding path is quite slow, but it doesn't run into the default gunicorn timeout of 30s (though it's almost there with its load time of ~20s).
However when I access the index page – the one listing all repositories – I get a 5xx error from nginx due to a gateway timeout.

I opened this one with klaus rather than dulwich because it only seems to get triggered by the index page, *not* the project specific page (and the only other project there has like 100 commits and has like a megabyte I guess).

# Details

Commands to reproduce:

```bash
# go somewhere with ~5GiB of diskspace
cd "$(mktemp -dp /tmp)"
# this takes ages
git clone --mirror https://github.com/NixOS/nixpkgs
# it did happen with other command lines but this is the last one in my history that I know worked
klaus * --port 8888 --debug

# this should run pretty fast (0.395s)
curl http://localhost:8888/nixpkgs/
# this will take very long (35.289s)
curl http://localhost:8888/
```

I can reproduce this on both Gentoo and NixOS (klaus 2.0.2, dulwich 0.21.3 and 0.20.50 respectively).
I can also reproduce it with the Docker image (`dd9eaa5d40c7`) using this:

```bash
podman run --rm -p 7777:80 -v $PWD/:/repos/ -it jonashaag/klaus:latest klaus --host 0.0.0.0 --port 80 /repos/nixpkgs.git
```

Note that sometimes I also get a traceback like the following after a minute or so of dulwich reading the pack files, the error seems spurious, no idea what it's about, it usually goes away after one or two occurrences:

<details>
<summary>zlib.error: Error -3 while decompressing data: incorrect header check</summary>

```text
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 2551, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/lib/python3.10/site-packages/klaus/utils.py", line 62, in __call__
    return super(ProxyFix, self).__call__(environ, start_response)
  File "/usr/lib/python3.10/site-packages/werkzeug/middleware/proxy_fix.py", line 187, in __call__
    return self.app(environ, start_response)
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 2531, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 2528, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/usr/lib/python3.10/site-packages/klaus/views.py", line 79, in repo_list
    repos = sorted(repos, key=sort_key)
  File "/usr/lib/python3.10/site-packages/klaus/views.py", line 75, in <lambda>
    -(repo.fast_get_last_updated_at() or -1),
  File "/usr/lib/python3.10/site-packages/klaus/repo.py", line 363, in fast_get_last_updated_at
    self.__last_updated_at = self.__repo.get_last_updated_at()
  File "/usr/lib/python3.10/site-packages/klaus/repo.py", line 65, in get_last_updated_at
    return cached_call(
  File "/usr/lib/python3.10/site-packages/klaus/repo.py", line 36, in cached_call
    data = producer()
  File "/usr/lib/python3.10/site-packages/klaus/repo.py", line 68, in <lambda>
    producer=lambda: self._get_last_updated_at(all_refs),
  File "/usr/lib/python3.10/site-packages/klaus/repo.py", line 75, in _get_last_updated_at
    resolveable_refs.append(self[ref_hash])
  File "/usr/lib/python3.10/site-packages/dulwich/repo.py", line 771, in __getitem__
    return self.object_store[self.refs[name]]
  File "/usr/lib/python3.10/site-packages/dulwich/object_store.py", line 123, in __getitem__
    type_num, uncomp = self.get_raw(sha1)
  File "/usr/lib/python3.10/site-packages/dulwich/object_store.py", line 541, in get_raw
    return pack.get_raw(sha)
  File "/usr/lib/python3.10/site-packages/dulwich/pack.py", line 2377, in get_raw
    obj_type, obj = self.data.get_object_at(offset)
  File "/usr/lib/python3.10/site-packages/dulwich/pack.py", line 1328, in get_object_at
    unpacked = self.get_unpacked_object_at(offset, include_comp=False)
  File "/usr/lib/python3.10/site-packages/dulwich/pack.py", line 1313, in get_unpacked_object_at
    unpacked, _ = unpack_object(self._file.read, include_comp=include_comp)
  File "/usr/lib/python3.10/site-packages/dulwich/pack.py", line 871, in unpack_object
    unused = read_zlib_chunks(
  File "/usr/lib/python3.10/site-packages/dulwich/pack.py", line 297, in read_zlib_chunks
    decomp = decomp_obj.decompress(add)
zlib.error: Error -3 while decompressing data: incorrect header check
```

</details>

### PR DESCRIPTION
**Pull request by jonashaag**
@benaryorg please try this, closes #309

### CODE REVIEW

File: klaus/repo.py [hunk 1]
```diff
@@ -58,10 +58,21 @@ def namespaced_name(self):
...
-        # Cache result to speed up repo_list.html template.
-        # If self.get_refs() has changed, we should invalidate the cache.
-        all_refs = self.get_refs()
+        """Get datetime of last commit to this repository.
+
+        Caches the result to speed up the repo_list page.
+        Cache is invalidated if one of the ref targets changes,
+        eg. a new commit has been made and 'refs/heads/master' was changed.
+        """
+        if len(self.refs.keys()) > 10000:
```
Comment (jelmer) [inline klaus/repo.py:67]: 10000 still seems like quite a lot, there's no way to meaningfully display that - and it still means 10k random accesses of the repository.

What if there are more than 10000 heads, as is the case with the nix repo?

Comment (jonashaag) [reply klaus/repo.py:67]: Ah damn I thought that most heads are branches but eg. GH PRs are also heads

Comment (jonashaag) [reply klaus/repo.py:67]: Note that the number used here is not displayed anywhere, just the maximum timestamp of the refs

**Comment by benaryorg** [comment]
> @benaryorg please try this

```text
curl -so /dev/null http://localhost:8000/  0.01s user 0.01s system 0% cpu 26.207 total max RSS 8
```

That's the initial request, still taking quite a long time (in fact almost hitting the default gunicorn timeout again).
After that of course it's cached and returns almost instantly.
The time I got with my quick&dirty version of the patch that only checked *HEAD* ran in ~25.514s which is barely any better, so I'd say this patch is as fast as this is going to get on that (admittedly slow) machine here.
So yeah, this solves my issue!
Thanks a lot!

---

For posterity's sake; the nixpkgs repository itself (which at this point already had its timestamp cached):

```text
curl -so /dev/null http://localhost:8000/nixpkgs/  0.01s user 0.01s system 0% cpu 2.610 total max RSS 8
```

File: klaus/repo.py (hunk 1 above)
Comment (jonashaag) [reply klaus/repo.py:67]: > Ah damn I thought that most heads are branches but eg. GH PRs are also heads

Wait no they are not

**Comment by jonashaag** [comment]
Hmm I wonder why it's still so slow. For me it takes around 1s. Is the repo on a slow hard drive?

@jelmer is there a way to batch-lookup timestamps with Dulwich that's faster?

**Comment by jelmer** [comment]
> Hmm I wonder why it's still so slow. For me it takes around 1s. Is the repo on a slow hard drive?
> 
> @jelmer is there a way to batch-lookup timestamps with Dulwich that's faster?

The main alternative is to iterate over most of the repository using something like Repo.object_store.iterobject_subset(). That will at least remove the need to reinflate delta bases more than once (which you would probably be doing when doing random access).

**Comment by jonashaag** [comment]
```py
next(repo.object_store.iterobjects_subset([repo[b"HEAD"].id]))
```

```
KeyError: b'5804b7fb5fa07ac9b3720a19aa207edfb1521133'
```

Am I using it wrong?

**Comment by jelmer** [comment]
> 



> ```python
> next(repo.object_store.iterobjects_subset([repo[b"HEAD"].id]))
> ```
> 
> ```
> KeyError: b'5804b7fb5fa07ac9b3720a19aa207edfb1521133'
> ```
> 
> Am I using it wrong?

That seems right, and works here (although on a different repo, obviously). Does repo[repo[b'HEAD'].id] work? What version of dulwich are you on?

**Comment by benaryorg** [comment]
> Hmm I wonder why it's still so slow. For me it takes around 1s. Is the repo on a slow hard drive?

A slow SSD, which should still be a lot faster than an HDD but not up to speed with current SSDs.
However these load times aren't IO bound but CPU bound.
The thread the request is running on *does* use up 100% of a single core on the machine during those 20s though, so it likely isn't impacted by any IO speeds.
Recalling the stacktrace I got at one point I'm pretty sure it's going through a lot of of zlib compressed data, though I can't say for sure what exactly takes so long.

**Comment by jonashaag** [comment]
> Repo.object_store.iterobject_subset()

Unfortunately it is actually slower:

```
In [11]: %timeit list(repo.object_store.iterobjects_subset(random.sample(list(refs.values()), 1000), allow_missing=True))
94.2 ms ± 2.46 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [12]: %timeit [repo[x] for x in random.sample(list(refs.values()), 1000)]
66.7 ms ± 2.5 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Comment by jonashaag** [comment]
@benaryorg what's your CPU? I wonder how it can be 20x slower than mine

**Comment by benaryorg** [comment]
> @benaryorg what's your CPU? I wonder how it can be 20x slower than mine

Intel(R) Atom(TM) CPU N2800   @ 1.86GHz

This is probably one of my oldest and slowest servers *\*sigh\**

**Pull request merged**